Introduce CUDA backend specification

AerialMantis opened the pull request:
Introduce a first draft of a CUDA backend specification covering a mapping of the platform, execution and memories models and interoperability with the CUDA API for SYCL implementations targeting a CUDA backend.

The backend specification has been added as an appendix to the SYCL 2020 specification.

A pull request for introducing a test plan to the SYCL CTS for the CUDA backend interoperability defined in this backend specification is here - https://github.com/KhronosGroup/SYCL-CTS/pull/207

Commit:
Added initial chapter skeleton for CUDA backend specification

Commit:
Added initial chapter skeleton for CUDA backend specification

npmiller committed:
Add introduction and mapping of context and queue for CUDA backend specification

npmiller committed:
Add introduction and mapping of context and queue for CUDA backend specification

t4c1 committed:
Add memory model section to CUDA backend specification

t4c1 committed:
Add memory model section to CUDA backend specification

Commit:
Add cuda execution model

Commit:
Add cuda execution model

Commit:
cuda extension

Commit:
cuda extension

JackAKirk committed:
Error handling

JackAKirk committed:
Error handling

JackAKirk committed:
Device and Platform

JackAKirk committed:
Device and Platform

npmiller committed:
Add API interoperability section

Commit:
Merge branch 'nicolas/interop' into 'SYCL-2020/cuda-backend-specification'

Add API interoperability section

See merge request oneapi-core/SYCL-Docs-mirror!7

npmiller committed:
Add API interoperability section

Commit:
Merge branch 'SYCL-2020/gordon/cuda-backend-specification-kernel-func-interop' into 'SYCL-2020/cuda-backend-specification'

Add CUDA backend specification kernel function interop definitions

See merge request oneapi-core/SYCL-Docs-mirror!10

Commit:
Add CUDA backend specification kernel function interop definitions

Commit:
Add CUDA backend specification kernel function interop definitions

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/SYCL-Docs?pullRequest=197) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/SYCL-Docs?pullRequest=197) before we can accept your contribution.<br/>**0** out of **5** committers have signed the CLA.<br/><br/>:x: Hugh Delaney<br/>:x: AerialMantis<br/>:x: npmiller<br/>:x: t4c1<br/>:x: JackAKirk<br/><hr/>**Hugh Delaney** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/SYCL-Docs?pullRequest=197) it.</sub>

npmiller commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+| [code]#template<backend Backend> +
+platform + 
+make_platform(const backend_input_t<Backend, platform> &backendObject);# 
+        | Create a SYCL `platform` from a list of CUDA device, the list must contain at least one CUDA device.
```
Only one platform makes sense but I'm not sure it's a requirement we want to introduce. For example at the moment DPC++ will return one platform per CUDA device, it will likely be changed in the future to only one platform, but still it's possible for vendors to organize their platforms in different ways.

It does make it a little strange if someone tries to create a platform with a subset of devices that's not in `get_platforms()`. So we might need to revisit this a bit.

gmlueck replied:
I think it is already a requirement because the SYCL core spec says the following in [section 4.6.2 "Platform class"](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:platform-class):

> The execution environment for a SYCL application has a fixed number of platforms which does not vary as the application executes. The application can get a list of all these platforms via platform::get_platforms(). The platform class also provides constructors, but constructing a new platform instance merely creates a new object that is a copy of one of the objects returned by platform::get_platforms().

npmiller replied:
Well yes, the fixed number of platforms is a requirement, what I meant was that I'm not sure the backend spec should mandate that all CUDA devices be under the same platform.

gmlueck replied:
Yes, I think that makes sense too.  I'd just like to see some wording in the description of `make_platform()` that the returned `platform` object is a copy of one of the objects returned by the core API `platform::get_platforms()`.  For example, this is what the Level Zero interop spec says:

> Constructs a SYCL platform instance from a Level-Zero `ze_driver_handle_t`. The SYCL execution environment contains a fixed number of platforms that are enumerated via `sycl::platform::get_platforms()`. Calling this function does not create a new platform. Rather it merely creates a `sycl::platform` object that is a copy of one of the platforms from that enumeration.

AerialMantis replied:
The wording for this has been updated to reflect that `make_platform` doesn't create a new platform just returns a `platform` object that is a copy of one of the existing platforms, so I think this is resolved now.

npmiller commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+
+When creating a SYCL object from a native CUDA object SYCL does not take
+ownership of the object and it is up to the application to dispose of them when
+appropriate.
```
Yes, we looked at that, and I do think it makes sense as well, however it does use an extension header, which is fine as the Level Zero is maintained separately, but it's not ideal if we want the CUDA backend spec to be part of the main SYCL specification.

So we decided to go with a simpler approach for now, for the CUDA backend spec, with the idea in mind that we should probably look into maybe codifying what Level Zero is doing directly in the SYCL interop API, or at least a similar system that could be re-used by the non-OpenCL backends to handle ownership.

gmlueck replied:
> however it does use an extension header, which is fine as the Level Zero is maintained separately, but it's not ideal if we want the CUDA backend spec to be part of the main SYCL specification.

Sorry, I'm not following.  What do you mean by "it does use an extension header"?  How does this preclude the interop API from being part of the main SYCL specification?

npmiller replied:
It doesn't preclude it, we just thought it might be better done separately as a generic change to the interop API that could be used by all backends.

gmlueck replied:
I'm not sure it makes sense to try to solve this in a general way in the core spec, since only some backends have this issue of "ownership".  (For example, OpenCL doesn't have this issue because objects are reference counted.)  For backends that have this issue (e.g. Level Zero, CUDA), I think it makes sense for each backend to define the mechanism for deciding who "owns" an object.  Of course, it would be better if all backends had a similar mechanism, though.

If the Level Zero mechanism seems good to others, can we use that for CUDA too?  If it has flaws, we could discuss them and potentially adopt a consistent mechanism across both backends.

npmiller replied:
I would argue that the OpenCL backend also has this issue of ownership, it's just that it happens to be easily solved by the OpenCL reference counting. And that the Level Zero approach could easily be implemented by the OpenCL backend, where the ownership transfer would just dictate how SYCL calls `clRetain*` and `clRelease*` on the OpenCL objects.

I could be completely wrong on that but looking at the interop API with `make_*` and `get_*` it seems to me that ownership questions would naturally come up and I get the feeling that maybe it wasn't part of the original interop API simply because it was solved so easily by reference counting for the OpenCL backend.

npmiller committed:
Fix editing typo and improve make_device docs

npmiller committed:
Fix editing typo and improve make_device docs

npmiller committed:
Merge branch 'nicolas/fixes' into 'SYCL-2020/cuda-backend-specification'

Fix editing typo and improve make_device docs

See merge request oneapi-core/SYCL-Docs-mirror!12

gmlueck commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+| [code]#context#  | `CUcontext`               | `std::vector<CUcontext>` | A SYCL context can encapsulate multiple CUDA contexts , however it is not possible to create a SYCL context from multiple CUDA contexts.
+| [code]#queue#    | `CUstream`   | `CUstream` | A SYCL queue encapsulates a CUDA stream.
+| [code]#event#    | `CUevent`    | `CUevent`  | A SYCL event encapsulates a CUDA event.
+| [code]#buffer# | `struct { CUdeviceptr ptr; size_t size; }` | `CUdeviceptr` | A SYCL buffer encapsulates a CUDA device pointer.
```
A question came up recently about the return value of `get_native()` for `buffer`.  What value is returned if `buffer` is a sub-buffer?  Is it a pointer to the original (non-sub) buffer, or is it an offset to the start of the sub-buffer?  It would be nice if there was a table describing the behavior of `get_native()` for each of these objects.  That table would be a good place to clarify things like this.

gmlueck replied:
Another question about the interop for `buffer` ... the OpenCL interop has a *vector* of `cl_mem` object because there might be multiple OpenCL devices on the system and the buffer might be represented on more than one device.  (At least I think that's the reasoning.)  Is it correct that there's just a single `CUdeviceptr` here, or should there be a vector of them?

AerialMantis replied:
So we've been discussing application interop for `buffer` as well and for a few reasons we came to the conclusion that we should not support it and instead encourage `buffer` interop via `interop_handle::get_native_mem`:
* As you mentioned, as a `buffer` can represent multiple devices it would need to return a vector of `CUdeviceptr`, but because it allocates data lazily, this would also mean we would either need to return a partial list of memory objects depending on the current state of the runtime or enforce allocation within the `get_native`.
* If the return value has multiple memory objects, we would need some way to reflect which one has the latest copy of the data.
* A `buffer` can also represent memory on the host and potentially across multiple backends, and representing this would add further complexity to the return type.

In effect we came to the conclusion we would need to replicate much of the inner workings of a `buffer` in order to make it really useful to users, and the use cases we had could be supported via host task interop.

AerialMantis commented:
@illuhad thanks for your further comments on this, apologies for the delay in getting back to you on this, I've been pulled away  to other things.

> I could of course just return some stream from the pool, but I'm not sure if this might not do more harm than good because I don't think it's what users would expect. In that case, there'd be no guarantees that the returned stream synchronizes in any way with previous operations submitted to the SYCL queue (they may have been scheduled to different streams, even for in-order queues) and neither would subsequent operations be guaranteed to synchronize with anything the user submits to the stream via interop.

So this is a really good point, and I think guarantees or expectations of synchronization need to be well defined. The way I've been looking at this is that if the user wants interop work to be integrated with other work enqueued by the SYCL queue, then they should use the host task interop as this is a much better tool for this (there are ways we need to improve the host task interop but I think that's a separate discussion). I see the get_native_queue API as a much more course grained mechanism for interop, for when users want to switch to using the native stream and stop using the SYCL queue while the native stream is being used, and as a result I think the synchronization rules can also be more course grained. Therefore I think we can state that get_native_queue returns a stream and it's up to the implementation which one, but require that the implementation performs any synchronization necessary such that if commands are enqueued to the native stream they would maintain consistency with commands enqueued to the SYCL queue prior to that. That may introduce some additional implementation burden, but a naïve implantation could always just synchronize the entire pool.

Another though that came up on this was that we may want to make the return type of get_native_queue a vector of CUDA streams, to allow for implementations which have a pool of streams or even 1:n streams for example if you have different streams for compute, data movement or interop, to reflect this.

> That might work. In most workflows, hipSYCL just uses the default context from the CUDA runtime API, which I think can be extracted if the device is known. There are some caveats since hipSYCL also supports multi-device queues, but those are not covered by the SYCL standard anyway.

I'm glad that should work for hipSYCL, the multi-device queue is an interesting feature, definitely something for us to think about in relation to the backend spec.

> I'm not sure those functions make much sense for hipSYCL. The hipSYCL interpretation of e.g. a queue is really more of a high-level interface to the SYCL task graph, with some functionality to manage synchronization of groups of tasks (e.g. wait()). To make this more clear I am pondering whether we should introduce the alias using task_group = queue; in hipSYCL. Relating a task group to a stream does not sound very useful semantically, and adding streams at runtime for interop purposes breaks some of the assumptions that the scheduler currently can make. I'll have to think some more about this. For context it might be even more complicated as we couldn't rely on the CUDA runtime API context anymore in the workflows that use it, and this in turn might break other interop use cases that currently "just work" for SYCL+CUDA interop applications that use the runtime API.

I agree, this definitely makes mapping an incoming native stream with a SYCL queue difficult. I thought of a further option for how we could implement this for a pool of streams implementation. If semantically the SYCL queue is simply an interface to the scheduler, a way to add commands to a global scheduler with information as to which device they are targeting, perhaps we could also raise the the relationship with a native stream to a similar level of abstraction. So when you call `make_queue` with a CUDA stream then it queries that stream to ensure there is a stream in the pool targeting the same device, synchronize with both the incoming native stream and the stream in the pool, and then associate the interop-created SYCL queue with that stream in the pool, so any work enqueued to the SYCL queue will always target that stream in the pool. This is technically not using the incoming stream object, but it should be equivalent to if did, whilst still utilising the streams from the pool.

> Sure, that could be implemented - assuming this is still useful for users then.

I've been thinking about this some more and I don't see any other way to define this. If we were to go this route then I think we may want to introduce a function for querying whether a SYCL event has a native event, such as `has_native_event` and then if this returns `false` then `get_native_event` will throw an exception.

> I wonder if maybe we should take a step back to see what we want to accomplish with the spec from a user's point of view to make sure that a common spec that targets the least common denominator between DPC++ and hipSYCL is really better for users than two separate implementation backend specs that are specifically built to match the DPC++ and hipSYCL designs?

So I think this is a really good point. From my perspective the motivation of having backend specs is to make the availability and semantics of backend interop APIs well-defined so that users can reason about them. The motivation for having a common backend spec between DPC++ and hipSYCL is to provide consistency and portability, so that users can write applications which include interop with the CUDA API and these can run across both implementations. Granted, due to the flexibility in the core specification to allow differences in implementations there does need to be a common denominator and this limits how much can be defined by the backend spec and therefore how much consistency can be guaranteed. However, I feel there is enough here to make this useful for users.

t4c1 commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+
+When accessors to images are created without [code]#target::host_buffer# they allocate memory, for example using [code]#cudaMalloc3DArray()# or [code]#cudaMallocArray()#. 
+
+When non-host accessors are created with [code]#target::host_buffer# they can, for example use [code]#cudaHostAlloc()# to allocate pinned memory on host.
```
Yeah, this is just wrong. Will be removed.

t4c1 commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+| device | [code]#cudaMalloc()#
+| host | [code]#cudaHostAlloc()#
+| shared | [code]#cudaMallocManaged()#
+|====
```
Agreed. The section on memory allocations will be removed.

t4c1 committed:
Update memory model

Commit:
Merge branch 'update_memory_model' into 'SYCL-2020/cuda-backend-specification'

Update memory model

See merge request oneapi-core/SYCL-Docs-mirror!11

Commit:
Merge branch 'SYCL-2020/cuda-backend-specification' of git.office.codeplay.com:oneapi-core/SYCL-Docs-mirror into SYCL-2020/cuda-backend-specification

Commit:
Updating cuda backend with notes on graphics APIs, builtin kernel functions,...

Commit:
Merge branch 'hugh/cuda-backend-spec' into 'SYCL-2020/cuda-backend-specification'

Updating cuda backend with notes on graphics APIs, builtin kernel functions,...

See merge request oneapi-core/SYCL-Docs-mirror!13

Commit:
Merge branch 'SYCL-2020/cuda-backend-specification' of git.office.codeplay.com:oneapi-core/SYCL-Docs-mirror into SYCL-2020/cuda-backend-specification

Commit:
Merge branch 'SYCL-2020/master' into SYCL-2020/cuda-backend-specification

npmiller committed:
Remove extra sentence and fix formatting in platform section

npmiller committed:
Merge branch 'platform' into 'SYCL-2020/cuda-backend-specification'

Remove extra sentence and fix formatting in platform section

See merge request oneapi-core/SYCL-Docs-mirror!15

npmiller committed:
Update make_platform wording to reflect fixed number of platfoms

npmiller committed:
Merge branch 'nicolas/platform-wording' into 'SYCL-2020/cuda-backend-specification'

Update make_platform wording to reflect fixed number of platfoms

See merge request oneapi-core/SYCL-Docs-mirror!16

npmiller committed:
Clarify interop return value for sub-buffers

npmiller committed:
Merge branch 'nicolas/sub-buffers' into 'SYCL-2020/cuda-backend-specification'

Clarify interop return value for sub-buffers

See merge request oneapi-core/SYCL-Docs-mirror!17

hdelan commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+The table should be developed with other members of the SYCL community.
+
+[[table.extensionsupport]]
+.SYCL support for CUDA 11.3 extensions
```
I think the idea of adding a column for feature test macros is a good idea. We want to discourage the use of sycl::aspects to determine the availability of extensions, since the extensions need to be configured at compile time rather than runtime. Therefore we will replace the SYCL Aspect column with a feature test macro column.

We can also remove the Core SYCL API column

hdelan replied:
sycl::aspects describing extensions are useful to check if a particular device has a capability at runtime. Do you think any mention of such aspects is outside the scope of this document?

AerialMantis replied:
We've updated the table to reflect the associated SYCL aspect and feature macro for each extension, the table is currently empty, but serves as a placeholder for us to populate in the future.

AidanBeltonS commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+`sycl::cuda::surface` respectively.
+
+Below is a table of the `backend_input_t` and `backend_return_t` specializations
+for the SYCL classes which support kernel function interoperability.
```
Agreed, this is a mistake. There is only `get_native` within the kernels. The `backend_input_t` should be removed from the kernel function interoperability.

AerialMantis replied:
This wording has been removed, so I think this is resolved now.

npmiller committed:
Remove buffer interop for CUDA backend

After discussions we've decided to go against buffer interop, host tasks
and accessor interop should be used instead.

hdelan commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+Additional CUDA features are available depending upon the devices compute capability.
+SYCL can support these optional CUDA features with extensions.
+Unlike OpenCL, CUDA needs to know if the extension is available at compile time. 
+As a result there are no valid CUDA extensions which can be passed to `has_extension`.
```
Agreed. This will be corrected in next version of this document.

AerialMantis replied:
This wording has been removed, so I think this is resolved now.

hdelan commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+Unlike OpenCL, CUDA needs to know if the extension is available at compile time. 
+As a result there are no valid CUDA extensions which can be passed to `has_extension`.
+
+As the extension must be known at runtime CUDA extensions are best implemented 
```
Correct thanks.

AerialMantis replied:
This wording has been removed, so I think this is resolved now.

hdelan commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+SYCL_EXT_<vendorstring>_<featurename>. For CUDA extensions this format translates 
+to SYCL_EXT_NVIDIA_<featurename>. Similarly, the format for the naming of extension 
+classes and enumerations should be ext_<vendorstring>_<featurename>. Which in this context
+becomes ext_NVIDIA_<featurename>. Given the necessity to know the extension at 
```
Internally we have decided that using the feature test macro format `SYCL_EXT_<vendor>_<feature>` should use the usual vendor (ONEAPI, HIPSYCL etc) but should also include the word `CUDA` if it is related solely to the CUDA backend. Therefore a feature test macro for some CUDA extension in oneAPI becomes `SYCL_EXT_ONEAPI_CUDA_<feature>`

hdelan replied:
This would make it easier to introduce extensions, as using `KHR` in the macro would require a vote from the entire SYCL group

hdelan commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+5.3 or greater.
+
+[[sub:cuda:extensions]]
+=== Extensions
```
This is a placeholder section for later extensions

AerialMantis replied:
This wording has been updated, so I think this is resolved now.

hdelan commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+[code]#half4#, [code]#half8# and [code]#half16# must be
+available at compile-time. However a kernel using these types is only
+supported on devices that have [code]#aspect::fp16#, i.e. compute capability
+5.3 or greater.
```
This is correct. We have rephrased this section in the next iteration.

AerialMantis replied:
This wording has been updated, so I think this is resolved now.

hdelan commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+are non-core CUDA features. Therefore, the runtime must be able to determine what aspects CUDA 
+devices have. This can be performed by querying `cudaDeviceProp::major` and `cudaDeviceProp::minor`
+to find out the compute capability. The compute capability indicates what extensions are
+available to the device, and therefore what aspects are available.
```
This is a valid point. We have omitted/rephrased some things in the latest iteration, which we will be posting shortly.

Commit:
Responding to comments for non-core features

Commit:
Merge branch 'hugh/cuda-backend-spec' into 'SYCL-2020/cuda-backend-specification'

Responding to comments for non-core features

See merge request oneapi-core/SYCL-Docs-mirror!19

Commit:
Merge branch 'no-buffer-interop' into 'SYCL-2020/cuda-backend-specification'

Remove buffer interop for CUDA backend

See merge request oneapi-core/SYCL-Docs-mirror!20

Commit:
Merge branch 'update_with_queries' into 'SYCL-2020/cuda-backend-specification'

Update CUDA backend specification with the information on queries.

See merge request oneapi-core/SYCL-Docs-mirror!14

t4c1 committed:
Update CUDA backend specification with the information on queries.

Commit:
Update kernel interop based on feedback

Commit:
Merge branch 'aidan/update-from-feedback' into 'SYCL-2020/cuda-backend-specification'

Update kernel interop based on feedback

See merge request oneapi-core/SYCL-Docs-mirror!18

Commit:
Merge branch 'gordon/cuda-backend-spec-edits' into 'SYCL-2020/cuda-backend-specification'

Make edits to backend spec based on feedback.

See merge request oneapi-core/SYCL-Docs-mirror!22

Commit:
Make edits to backend spec based on feedback.

AerialMantis commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+| [code]#template<backend Backend> +
+device +
+make_device(const backend_input_t<Backend, device> &backendObject);# 
+        | Create a SYCL `device` from a CUDA device.
```
The wording for this has been updated to reflect that `make_device` doesn't create a new platform just returns a `device` object that is a copy of one of the existing devices, so I think this is resolved now.

AerialMantis commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+| [code]#unsampled_image_accessor<T, 1, Mode, image_target::device># | sycl::cuda::surface<T, 1> | sycl::cuda::surface<T, 1> | Convert a SYCL `accessor` to the `sycl::cuda::surface` interoperability type with the same type and dimensions.
+| [code]#unsampled_image_accessor<T, 2, Mode, image_target::device># | sycl::cuda::surface<T, 2> | sycl::cuda::surface<T, 1> | Convert a SYCL `accessor` to the `sycl::cuda::surface` interoperability type with the same type and dimensions.
+| [code]#unsampled_image_accessor<T, 3, Mode, image_target::device># | sycl::cuda::surface<T, 3> | sycl::cuda::surface<T, 1> | Convert a SYCL `accessor` to the `sycl::cuda::surface` interoperability type with the same type and dimensions.
+| [code]#stream#                                                     | signed char * | signed char * | Convert a SYCL `accessor` to an undecorated raw signed char pointer.
```
Interop for the `stream` class has been removed, so I think this is resolved now.

AerialMantis commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+
+[[table.cuda.kernelinterop.nativeobjects]]
+.Types of native backend objects kernel function interoperability
+[width="100%",options="header",cols="20%,20%,20%,40%"]
```
Yes this looks like a copy/paste error, I've corrected this.

AerialMantis commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,496 @@
+[[sub:cuda:platform_model]]
+=== Platform Model
+
+All CUDA enabled devices which can be executed on are represented by a single `CUdevice`. A SYCL device maps to a single CUDA device.
```
This wording has been updated to address this, so I think this is resolved now.

AerialMantis commented:
@illuhad I've tried to address the points you raised in recent changes.

I've added clarification to the SYCL `queue` interop to reflect that the SYCL queue can represent multiple CUstream objects however a SYCL `queue` will only produce or be created from one CUstream, and that the SYCL runtime should perform any necessary synchronization when transitioning between a SYCL `queue` and a CUstream.

I've introduced the backend function `sycl::cuda::has_native_event` and updated the wording for the SYCL `event` interop to handle the case where an internal CUevent is elided and therefore no valid CUevent can be returned.

fraggamuffin commented:
request more feedback

fraggamuffin commented:
need more reviews, but may be after IWOCL/SYCLCon

AerialMantis committed:
Update adoc/chapters/cuda_backend.adoc

Co-authored-by: Jakub Chlanda <j.chlanda@gmail.com>

AerialMantis committed:
Update adoc/chapters/cuda_backend.adoc

Co-authored-by: Jakub Chlanda <j.chlanda@gmail.com>

AerialMantis committed:
Update adoc/chapters/cuda_backend.adoc

Co-authored-by: Jakub Chlanda <j.chlanda@gmail.com>

AerialMantis committed:
Update adoc/chapters/cuda_backend.adoc

Co-authored-by: Jakub Chlanda <j.chlanda@gmail.com>

gmlueck commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+| [code]#device#   | `CUdevice`                | `CUdevice`               | A SYCL device encapsulates a CUDA device.
+| [code]#context#  | `CUcontext`               | `std::vector<CUcontext>` | A SYCL context can encapsulate multiple CUDA contexts, however, it is not possible to create a SYCL context from multiple CUDA contexts.
+| [code]#queue#    | `CUstream`   | `CUstream` | A SYCL queue can encapsulates multiple CUDA stream, however, a SYCL queue can only be created from or produce one, and any synchronization required should be performed.
+| [code]#event#    | `CUevent`    | `CUevent`  | A SYCL event can encapsulates multiple CUDA events, however, a SYCL event can only be created from or produce one, and a CUevent produced from a SYCL event may or may not be valid, use `sycl::cuda::has_native_event` to query this.
```
What happens if you call `get_native` for an `event` that  does not have a native event?

Same question here as above for `queue`.  What does it mean that a SYCL `event` could contain more than one native even but only one is returned from `get_native`?  Is it important for the application to know which is returned?

JackAKirk replied:
cc @AerialMantis

gmlueck replied:
For OpenCL, `get_native<event>` returns a vector of `cl_event` objects.  Why wouldn't we do the same thing for CUDA and return a vector of `CUevent` objects?

AerialMantis replied:
So that is definitely an option, we did think about having `get_native<event>` return all native events that are associated with the SYCL `event`, however, it introduces the problem of conveying to the user what each of these events represent in order for the user to utilise the individual events. Whereas we believe generally users just wish to just synchronize with all the commands associated with the `event`. We could expand this in the future to return all native events, but we thought that it was best to keep it simple to begin with until we have some more user experience with it. If you think this would be the better approach we could reconsider it.

gmlueck replied:
I also don't know what is easiest for users.  I just seems weird that we would decide one thing for OpenCL and a different thing for CUDA.  It will probably be difficult to change this later to return a vector of events because that would be an API-breaking change.

gmlueck commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+| [code]#bool sycl::cuda::has_native_event(sycl::event)# | Returns `true` if the SYCL event has a valid `CUevent` that can be queries via application interop.
+|====
+
+Additional CUDA features are available depending upon the device's compute 
+capability. SYCL can support these optional CUDA features with extensions.
+
+Use of CUDA extensions requires that the API for a given extension is available
+to the SYCL implementation. This needs to be determined at compile time. 
+Checking for the existence of feature test macros is the preferred method
+for checking whether an API exists. The feature test macro format 
+is `SYCL_EXT_<vendor>_<feature>`. The `<vendor>` string may also contain the
+word `CUDA` for features specific to CUDA. For example, the feature test macro
+for CUDA extensions in oneAPI may be either `SYCL_EXT_ONEAPI_CUDA_<feature>`,
+or just `SYCL_EXT_ONEAPI_<feature>`.
+
+Use of a given CUDA extension also requires that a chosen device has the
+required compute capability to use the CUDA extension. This can be determined
+using `sycl::aspect`s. Non-core SYCL aspects may be defined by an
+implementation which would allow this check to happen at runtime.
+
+The table below shows a proposal for SYCL supported CUDA extensions. This should
+be populated by other members of the SYCL community.
+
+[[table.extensionsupport]]
+.SYCL support for CUDA 11.3 extensions
+[width="100%",options="header",cols="35%,35%,15%, 15"]
+|====
+| CUDA Extension                            | SYCL Aspect   | Feature Test Macro               | Required Compute Capability 
+|====
```
I don't think any of this needs to be in the interop specification.  Vendors are always free to publish extensions, and each extension should describe its interface such as macros, aspects, etc.  I don't think we need any general description of this in the core backend interop specification.

I also don't think we want a list of extensions here in the core backend interop specification.  We don't want to update this backend interop spec (and get Khronos ratification) each time a vendor publishes an extension.

JackAKirk replied:
Sounds sensible to me. @AerialMantis ?

JackAKirk replied:
OK I will remove this.

gmlueck commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+codes may originate in the SYCL runtime or be passed from other runtimes to
+the SYCL runtime. When a `sycl::exception` is thrown, the `sycl::errc` can
+be queried using the exception's `.code()` method. Possible values for 
+`sycl::errc` include: `success`, `runtime`, `memory_allocation`, and more.
```
I wonder if we want this last sentence.  It seems to place an unnecessary dependency between this interop spec and the core SYCL spec.  Is it important to list the specific `sycl::errc` error codes?

JackAKirk replied:
I also think it is not important to list the specific `sycl::errc` codes in this way and the sentence can be removed.

JackAKirk replied:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

gmlueck commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+`sycl::errc` include: `success`, `runtime`, `memory_allocation`, and more.
+
+If there is a CUDA driver API error associated with an exception triggered, then the
+CUDA error code can be obtained by the free function `CUresult sycl::cuda::get_error_code(sycl::exception&)`. In the case where there is
```
```suggestion
CUDA error code can be obtained by the free function `CUresult sycl::cuda::get_error_code(const sycl::exception&)`. In the case where there is
```

We probably want this to be `const`?

JackAKirk replied:
Thanks for pointing this out. I don't see a reason why it should not be const, although I also don't see a reason why the OpenCL backend equivalent (Which this CUDA backend "Error handling" section tried to match) is not const: See SYCL 2020 rev 4 page 489

**_C.7.4. Errors and limitations**
If there is an OpenCL error associated with an exception triggered, then the OpenCL error code can be
obtained by the free function `cl_int sycl::opencl::get_error_code(sycl::exception&)`. In the case where
there is no OpenCL error associated with the exception triggered, the OpenCL error code will be CL_SUC­
CESS._

I think that both backends specs should match and be `const` unless there is a good reason otherwise: in which case this reason should be described in the spec. If there is a good reason that the OpenCL backend case is not `const` then for the time being we could mention this reason here as an answer to why we use `const` here, although ideally the OpenCL backend equivalent will be immediately updated and we simply make the change you suggest which appears obvious.

JackAKirk replied:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

gmlueck commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+no CUDA error associated with the exception triggered, the CUDA error
+code will be `CUDA_SUCCESS`.
+
+Most of the SYCL error codes that form sycl::errc are specifically defined as errors thrown during calls to the SYCL API or SYCL runtime. There are also some cases of sycl::errc which cover errors thrown during the compilation or execution of device code.
+It is suitable to map CUDA errors to such cases, such that an exception, "cuda_exception", that was created due to a CUDA error, may, upon execution of `cuda_exception.code()`, return a `std::error_code` relating to the `sycl::errc` case that the CUDA error maps to; whilst `sycl::cuda::get_error_code(cuda_exception)` will return the original CUDA error code.
+
+The relevant `sycl::errc` cases and the CUDA errors that they may be mapped from are listed below.
+
+==== build
+
+`sycl::errc::build` is defined as:
+
+_Error from an online compile or
+link operation when compiling,
+linking, or building a kernel bundle for a device._
+
+which may be mapped from `CUDA_ERROR_NO_BINARY_FOR_GPU`, `CUDA_ERROR_JIT_COMPILER_NOT_FOUND`, `CUDA_ERROR_INVALID_PTX`, `CUDA_ERROR_UNSUPPORTED_PTX_VERSION`, `CUDA_ERROR_SHARED_OBJECT_INIT_FAILED`, `CUDA_ERROR_SHARED_OBJECT_SYMBOL_NOT_FOUND`.
+
+==== memory_allocation
+
+`sycl::errc::memory_allocation` is defined as:
+
+_Error on memory allocation on the
+SYCL device for a SYCL kernel._
+
+which may be mapped from `CUDA_ERROR_OUT_OF_MEMORY`.
+
+==== kernel_argument
+
+`sycl::errc::kernel_argument` is defined as:
+
+_The application has passed an invalid argument to a SYCL kernel
+function. This includes captured
+variables if the SYCL kernel function is a lambda function._
+
+which may be mapped from `CUDA_ERROR_NOT_FOUND`.
```
I wonder if we want this part, though.  Do we really want to list the specific CUDA error codes that could correlate to each SYCL `errc`?  For example, do you want the CTS to fail if `cuda::get_error_code()` returns some CUDA error that is *not* in this list?  I think it might be better to leave this mapping as an implementation detail.

gmlueck replied:
There is probably a larger issue here about which error codes should be listed in the SYCL specification.

One option is that the spec should only list error codes that are specified to be thrown someplace in the core spec.  For example, `errc::build` is one of the error codes listed for the kernel bundle APIs, so the core spec needs to list `build` as one of the error codes.  By this rationale, we would remove any error codes that are not specifically called out anywhere in the spec.  (This is the topic of internal issue 512.)

For backend-specific errors, we would throw a SYCL exception with some generic SYCL error code (e.g. `errc::runtime`), and then it would be up to the application to use `sycl::cuda::get_error_code()` in order to get the CUDA specific error.  This way, we do not need to worry about which SYCL error corresponds to each backend error, and we don't need to add new SYCL errors whenever a new backend has an error that doesn't map well.

JackAKirk replied:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

gmlueck commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%% begin cuda_backend %%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Putting some general formatting comments here just so the conversation can be threaded:

* The SYCL spec uses `[code]#...#` for code markup not backticks.  All of the backticks should be changed to use `[code]#...#` instead.

* I got the following warnings when building the Asciiidoc sources:
  
  ```
  asciidoctor: WARNING: chapters/cuda_backend.adoc: line 481: id assigned to block already in use: table.extensionsupport
  asciidoctor: WARNING: chapters/cuda_backend.adoc: line 523: id assigned to block already in use: table.extensionsupport
  ```

* The formatting of some of the tables could be improved, but you can only see this in the rendered PDF or HTML version.  I find it easiest to install the Docker container, which allows you to build these yourself.  Another option is to download the "spec-outputs" artifact from GitHub.  (Click on the "Checks" tab in this PR.)

JackAKirk replied:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+|====
+| Feature                                   | SYCL Aspect               | Required Compute Capability 
+| [code]#16-bit floating point#             | [code]#aspect::fp16#      | 5.3 or greater
+|====
```
I can personally see how it isn't strictly necessary, however IMO it is quite concise and could be useful to have this table since the compute capability requirement information is not always easy to find out and it seems nice for people to know the requirement before they explicitly try to write code calling the aspect. @hdelan what do you think?

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+Other kernel function types in CUDA are represented by aliases provided in the
+`sycl::cuda` namespace. These are provided for the `sampled_image_accessor`,
+and `unsampled_image_accessor` classes; `sycl::cuda::texture` and
+`sycl::cuda::surface` respectively.
```
@AerialMantis what do you think?

AerialMantis replied:
That's a good point, we should specify what types these are aliases to. There's actually two options for this as CUDA has both the texture and surface reference API (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#texture-reference-api, https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#surface-reference-api) and the texture and surface object API (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#texture-object-api, https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#surface-object-api), though the reference API is deprecated so I think the object API makes more sense, in which case the types would be `cudaTextureObject_t` and `cudaSurfaceObject_t`. The object API is actually not templated like the reference API is, so I think we can remove the template arguments on the aliases in the SYCL CUDA backend.

So it would be something like:

```
namespace sycl::cuda {

using texture = cudaTextureObject_t ;
using surface = cudaSurfaceObject_t ;

}  // namespace sycl::cuda
```

gmlueck replied:
Thanks for the explanation, @AerialMantis.  I still don't understand why we have aliases for these two types, though.  Why can't `backend_return_t` simply be an alias directly to `cudaTextureObject_t`.  Why does it help to have the intermediate alias named `sycl::cuda::texture`?

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+| [code]#platform# | `std::vector<CUdevice>`   | `std::vector<CUdevice>`  | A SYCL platform encapsulates a list of CUDA devices.
+| [code]#device#   | `CUdevice`                | `CUdevice`               | A SYCL device encapsulates a CUDA device.
+| [code]#context#  | `CUcontext`               | `std::vector<CUcontext>` | A SYCL context can encapsulate multiple CUDA contexts, however, it is not possible to create a SYCL context from multiple CUDA contexts.
+| [code]#queue#    | `CUstream`   | `CUstream` | A SYCL queue can encapsulates multiple CUDA stream, however, a SYCL queue can only be created from or produce one, and any synchronization required should be performed.
```
@npmiller @AerialMantis what do you think?

gmlueck replied:
Thanks, this seems reasonable.  If we add a table describing the `get_native` functions, the description for `get_native<queue>` can explain that the implementation may use several `CUstream` objects for each `queue`, but `get_native` returns just one and the choice is implementation defined.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+
+For all event information profiling descriptors, the calls to 
+[code]#sycl::event::get_profiling_info# return the time (in nanoseconds) since 
+the creation of the context that the event is associated with. The "Resolution"
```
@t4c1 what do you think?

BTW should the meaning of "timebase" be obvious? The interpretation for it that I get from #242 is that it represents "timestamp" from an event that has the same invariant value across a given backend. I thought this should also be satisfied in the CUDA backend with the caveat of the large 10^-6 error range. BTW do other backends have such large errors with respect to the nanosecond unit time? Should a note that the error can be much larger than 10^-9 in some backends be added to the content from #242?

t4c1 replied:
The DPC++ implementation can be changed to use a common base event, for example, in the platform object. So it can conform to the changes in the SYCL spec, but the CUDA backend spec needs to be updated.

gmlueck replied:
I don't think the spec needs to say anything about the granularity of the timestamps.  This seems like a "quality of implementation" thing.

gmlueck replied:
I think it is OK for this CUDA backend specification to note the timer resolution, so  it is OK to keep this wording:

> The resolution (timing error) of the returned value is the same as that provided by the CUDA driver API call `cuEventElapsedTime`: +/- 0.5 microseconds.

Does everyone agree that this statement (and the implementation) will change, though?

> For all event information profiling descriptors, the calls to  `sycl::event::get_profiling_info` return the time (in nanoseconds) **since the creation of the context** that the event is associated with.

Instead, we want all profiling timestamps from the CUDA backend to share the same "timebase".  It sounds like you think this is possible to implement, @t4c1?

t4c1 replied:
Yes, it should not be too much work.

JackAKirk replied:
@t4c1 can you specify what the exact wording of the update should be now?

t4c1 replied:
Something like: 
```suggestion
the creation of the CUDA platform (which happens when application is started). The "Resolution"
```

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+
+It is best to avoid calculating the linear index manually; it is better
+to use a multi-dimensional `sycl::id` to index into memory, as it doesn't
+expose index-flipping to the user.
```
@hdelan I also find it a bit unclear in parts. Perhaps you can clarify a little that the CUDA backend uses a different format to the SYCL frontend (if this is necessary/useful to mention here?) and best practice for a user to avoid running into related issues/ what they need to know about the index flipping if they can't avoid running into this issue?

JackAKirk replied:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+unit. CUDA uses a hierarchy of threads to organize the execution of
+kernels. Kernels are split up into thread blocks. The threadblocks 
+form a grid each thread can identify its location within the grid 
+using a block ID. The grid is a concept used to index threadblocks 
+the grid can be one, two, or three dimensions. Each thread block is 
```
+1

JackAKirk replied:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+
+==== Atomics
+
+Not all CUDA devices support all memory orders. If a particular memory order is unsupported by a CUDA device, it can be unsupported in the SYCL CUDA backend for that device. Sequentially consistent atomics are currently not supported on any device, so the SYCL CUDA backend is not required to implement them. The mappings of other memory orders (when supported by the device) is defined in table <<t ...
```
+1 for removing the line: I thought that relaxed meant no restrictions so extra limitations wouldn't be possible by definition? In any case the most basic memory order in CUDA is relaxed with a matching definition to the SYCL spec.

t4c1 replied:
I put that here to emphasize that support for memory orders depends on device, not on the implementation of the backend. It could as well be removed.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+[width="100%",options="header",cols="40%,30%,30%"]
+|====
+| SYCL shared USM advice | CUDA managed memory advice | processor, the advice is set for
+| CUDA_MEM_ADVISE_SET_READ_MOSTLY | cudaMemAdviseSetReadMostly | device associated with the queue/handler
```
@AidanBeltonS what do you think?

t4c1 replied:
I agree - this was an oversight. These should be lowercase and in sycl::cuda namespace.

JackAKirk replied:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+CUDA contexts require to be activated on the current thread to be used by other
+CUDA entry points. Therefore any use of the SYCL APIs with a CUDA backend may
+modify the current active context on the thread, and no guarantee is provided
+that any existing active CUDA context would be restored by SYCL.
```
cc @npmiller ^

npmiller replied:
So for the context part you are understanding that correctly, and yes it is legal for the application to change the active context before the next SYCL API call. Essentially the SYCL runtime must ensure that it's using the correct context associated with the `sycl::context`, but we don't want to force it to have to restore existing CUDA contexts, as this is a concern for performance and can sometimes lead to errors as well. This stems from issues we had in DPC++ (see write-up here: https://github.com/intel/llvm/pull/4442 ), and trying to keep the guidelines as general as possible so they can work with different implementations.

About the active devices, this is a very good question, so CUDA has two different APIs, the "driver API" which is lower level and used in DPC++ and the "runtime API" which is higher level and used for example in hipSYCL. The concept of active device is a "runtime API" concept and it's not in the "driver API", but it would make sense to touch on that in this CUDA backend spec as well.

I had another look through the CUDA Runtime and Driver APIs and as far as I can tell devices and contexts are the only objects to have this concept of being "active".

I'll have a second look at this and the `device` description to try and clarify this a bit and maybe make it more friendly to the runtime API as well.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+[width="100%",options="header",cols="20%,20%,20%,40%"]
+|====
+| [code]#SyclType# | [code]#backend_input_t<backend::cuda, SyclType># | [code]#backend_return_t<backend::cuda, SyclType># | Description
+| [code]#platform# | `std::vector<CUdevice>`   | `std::vector<CUdevice>`  | A SYCL platform encapsulates a list of CUDA devices.
```
cc @npmiller ^^

npmiller replied:
> For example, suppose there are two CUDA devices D1 and D2. Calling platform::get_platforms() returns one CUDA platform with those two devices. What happens if the application calls make_platform<backend::cuda>({D1}) (i.e. only provides one of the devices)?

I think in that case I would expect `make_platform` to return an error and require both devices.

But you're right that the platform interop doesn't really make sense for CUDA, I think it's just better to remove it.

JackAKirk replied:
OK I can remove the platform interop as suggested.

JackAKirk replied:
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+| [code]#SyclType# | [code]#backend_input_t<backend::cuda, SyclType># | [code]#backend_return_t<backend::cuda, SyclType># | Description
+| [code]#platform# | `std::vector<CUdevice>`   | `std::vector<CUdevice>`  | A SYCL platform encapsulates a list of CUDA devices.
+| [code]#device#   | `CUdevice`                | `CUdevice`               | A SYCL device encapsulates a CUDA device.
+| [code]#context#  | `CUcontext`               | `std::vector<CUcontext>` | A SYCL context can encapsulate multiple CUDA contexts, however, it is not possible to create a SYCL context from multiple CUDA contexts.
```
cc @AerialMantis ^

npmiller committed:
Use device version to report compute capabilities

JackAKirk committed:
Error Handling more concise

Signed-off-by: JackAKirk <jack.kirk@codeplay.com>

JackAKirk committed:
Clarified mem consist model and formating

Signed-off-by: JackAKirk <jack.kirk@codeplay.com>

npmiller committed:
Remove platform interop

Platform is not a concept that makes sense for CUDA and are simply a
list of devices so specific interop for it doesn't really make sense.

The user can already access the list of SYCL device then use interop on
these if necessary.

JackAKirk committed:
fixed __ formatting in __syncthreads

Signed-off-by: JackAKirk <jack.kirk@codeplay.com>

JackAKirk committed:
clarify backend spec is for `backend::cuda`

Signed-off-by: JackAKirk <jack.kirk@codeplay.com>

npmiller committed:
Merge branch 'nicolas/rm-platform' into 'SYCL-2020/cuda-backend-specification'

Remove platform interop

See merge request oneapi-core/SYCL-Docs-mirror!25

Commit:
Updating section on index flipping

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+
+[[table.noncorefeatures]]
+.CUDA support for non-Core SYCL APIs
+[width="100%",options="header",cols="33%,33%,33%"]
```
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+
+[[table.cuda.kernelinterop.nativeobjects]]
+.Types of native backend objects kernel function interoperability
+[width="100%",options="header",cols="20%,20%,20%,40%"]
```
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+single thread is a SYCL work-item.
+
+CUDA primarily synchronizes the threads through two functions,
+`cudaStreamSynchronize()` and `__syncthreads()`. 
```
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+This chapter describes how the SYCL general programming model is mapped on top
+of CUDA, and how the SYCL generic interoperability interface must be
+implemented by vendors providing SYCL for CUDA implementations to ensure SYCL
+applications written for the CUDA backend are interoperable.
```
This has been dealt with in our internal draft, commit here will be made once other points are dealt with.

JackAKirk committed:
Removed platform interop.

Signed-off-by: JackAKirk <jack.kirk@codeplay.com>

JackAKirk committed:
Updated timing descriptor ref to platform creation time.

Signed-off-by: JackAKirk <jack.kirk@codeplay.com>

npmiller committed:
Clarify active contexts and devices

Commit:
Remove is_backend_active

Commit:
Merge branch 'nicolas/fix-version' into 'SYCL-2020/cuda-backend-specification'

Use device version to report compute capabilities

See merge request oneapi-core/SYCL-Docs-mirror!23

Commit:
Merge branch 'nicolas/clarify-context' into 'SYCL-2020/cuda-backend-specification'

Clarify active contexts and devices

See merge request oneapi-core/SYCL-Docs-mirror!28

Commit:
Merge branch 'update_cuda_spec' into 'SYCL-2020/cuda-backend-specification'

Remove is_backend_active

See merge request oneapi-core/SYCL-Docs-mirror!29

Commit:
Merge branch 'SYCL-2020/cuda-backend-specification' into 'updates'

# Conflicts:
#   adoc/chapters/cuda_backend.adoc

Commit:
Merge branch 'SYCL-2020/update-index-flipping' into 'SYCL-2020/cuda-backend-specification'

Updating section on index flipping

See merge request oneapi-core/SYCL-Docs-mirror!27

Commit:
Merge branch 'SYCL-2020/cuda-backend-specification' into 'updates'

# Conflicts:
#   adoc/chapters/cuda_backend.adoc

JackAKirk committed:
Merge branch 'updates' into 'SYCL-2020/cuda-backend-specification'

updated MR for cuda backend spec

See merge request oneapi-core/SYCL-Docs-mirror!31

AerialMantis commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,598 @@
+context +
+make_context(const backend_input_t<Backend, context> &backendObject,
+                     const async_handler asyncHandler = {});# 
+        | Create a SYCL `context` from a CUDA context.
```
@illuhad you make a good point, thanks for raising this. I think we could resolve this by clarifying the wording such that for some SYCL implementations the mapping of SYCL context to CUDA context is limited to the CUDA primary context, which would limit the usefulness of the feature when the implementation is based on the runtime API, however, would allow us to maintain a generic interface between the two implementation approaches.

So the wording would be clarified such that `get_native<context, backend::cuda>` is permitted to return the CUDA primary context and `make_context<backend::cuda>` is permitted to only accept the CUDA primary context, and throw an exception otherwise.

However, in order to allow users to reason about this there would need to be some kind of query which exposes whether the implementation supports interop with non-primary CUDA contexts. Aside from this particular use case, there is another situation where having such a query would be important. Within a host task, when interoperating with the native objects, it's important to know whether the SYCL implementation initializes non-primary contexts, in order to ensure that the native API calls within the host task are compatible.

I see a couple of ways we could do this:
* A. Introduce a query which tells the user specifically whether the SYCL implementation maps all SYCL contexts to the CUDA primary context or whether unique contexts are allocated, this provides the specific information we need in this case, but if perhaps less scalable.
* B. Introduce a broader query which tells the user whether the  SYCL implementation is using the driver API or the runtime API, this could be useful for other places where there is a difference between the drive API and the runtime API, though it may not reflect the case where an implementation combines the two APIs.

illuhad replied:
Thanks for your thoughts @AerialMantis! I agree that what you are proposing are possible solutions. I have a hunch that in practice it might be easier to work with for users if they know if an implementation is runtime or driver API based rather than make the query more fine-grained and just about context interop.
E.g. if you know that your CUDA backend is runtime API-based, you can make other assumptions about backend interop. In that case, you can know that backend interop is generally just going to work without having to worry about contexts at all, which I think can be very valuable.
I wonder if it might be easier to understand and clearer if the backend specification just had two flavors, or profiles, so that we can expose the behavior of both approaches in a more obvious way.

I suspect that even if an implementation were to utilize both APIs, you could probably figure out a definition to classify them, most likely around how they behave regarding context interop, and whether the current CUDA device from `cudaSetDevice()` plays a role.
For example, hipSYCL already uses some driver API components for manual management of PTX modules, but still follows the interop behavior of the runtime API-approach.
Do you think there are use cases where such a classification is not possible?

gmlueck replied:
> I wonder if it might be easier to understand and clearer if the backend specification just had two flavors, or profiles, so that we can expose the behavior of both approaches in a more obvious way.

Just throwing this out there ... it is also an option to have two different CUDA backends.  This means:

* There would be two different enumerators in `enum backend`.
* Each backend would be responsible for defining its own interop specification (if it wanted to support interop).

Having a single backend interop specification is nice, but only if developers can reasonably write interop code that is portable across both CUDA backends.  If we think the two CUDA backends are just too different, it might be easier for users if each one had its own separate interop specification.

AerialMantis replied:
@illuhad @gmlueck I think you make a good point, I am wary of introducing different profiles of the CUDA backend specification if possible, I think it would be valuable to have a single specification that all implementations can follow, though I recognize this may not be possible if there are cases we can't handle generically.

I've clarified the wording to state that both APIs are supported. The biggest concern there was regarding the context and whether the primary context is used. I've added a note that the CUcontexts that are returned from interop can include the primary context, and in fact, this is now what the DPC++ CUDA backend does. I suspect we may still need a query for whether the implementation is using the CUDA Driver API or CUDA Runtime API, but I'd like to avoid this if possible to avoid having any bifurcation of the specification.

> Also, I wonder: Do we believe that there will actually be a use case to invoke make_context with the same context obtained from get_native()? Or will having this query for runtime API backends just tempt users to do things that are unsupported?

This is a good point, I suspect we don't want to support this, it would like cause havoc with the ownership model.

fraggamuffin commented:
still need to discuss how to handle context interrupt

JackAKirk commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,551 @@
+[width="100%",options="header",cols="40%,30%,30%"]
+|====
+| SYCL shared USM advice | CUDA managed memory advice | processor, the advice is set for
+| [code]#sycl::cuda::advice::cuda_mem_advise_set_read_mostly# | [code]#cudaMemAdviseSetReadMostly# | device associated with the queue/handler
```
@AerialMantis FYI these `advise` names seem to be different than those used in DPC++: https://github.com/t4c1/llvm-test-suite/blob/1a5a4e9d72cccf290d88f2498c08314d8bfb271e/SYCL/USM/memadvise_cuda.cpp#L39

Do we need to update one or other of the documentation / dpc++ impl?

AerialMantis replied:
Good catch, these are different, but I think we may want to update DPC++ to match what we have in here. I have also updated the definitions here to follow the correct extension convention.

AerialMantis commented on adoc/chapters/cuda_backend.adoc:
```diff
@@ -0,0 +1,551 @@
+
+==== Samplers
+
+In both SYCL and CUDA samplers consist of addressing mode, filtering mode and coordinate normalization mode. Mapping between SYCL and CUDA values is defined in tables <<table.cuda.memmodel.sampler_addressing>>, <<table.cuda.memmodel.sampler_filtering>> and <<table.cuda.memmodel.sampler_normalization>>. In CUDA addressing modes for all dimesnions will be the same, as CUDA allows different addressi ...
```
Fixed.

AerialMantis commented:
@illuhad apologies that it took me so long to get back to this, I've created a significant update to the pull request and actually moved this to a new PR as I ran into issues rebasing the changes.

> I could of course just return some stream from the pool, but I'm not sure if this might not do more harm than good because I don't think it's what users would expect. In that case, there'd be no guarantees that the returned stream synchronizes in any way with previous operations submitted to the SYCL queue (they may have been scheduled to different streams, even for in-order queues) and neither would subsequent operations be guaranteed to synchronize with anything the user submits to the stream via interop.

Regarding your comment about which CUDA stream is returned when interoperating with a SYCL queue, and how to correctly synchronize with this, I clarified the wording such that both the application interop and host task interop require synchronization with the SYCL queue prior to retrieving the native CUDA stream or executing the host task. This is quite strict, and perhaps not the most efficient approach, however, it does ensure correctness, and I think we could address this at a later point with a general improvement to the core specification with a feature like your extension for custom operations or a similar extension that we've been working on at Codeplay.

> I'm not sure those functions make much sense for hipSYCL. The hipSYCL interpretation of e.g. a queue is really more of a high-level interface to the SYCL task graph, with some functionality to manage synchronization of groups of tasks (e.g. wait()). To make this more clear I am pondering whether we should introduce the alias using task_group = queue; in hipSYCL. Relating a task group to a stream does not sound very useful semantically, and adding streams at runtime for interop purposes breaks some of the assumptions that the scheduler currently can make. I'll have to think some more about this. For context it might be even more complicated as we couldn't rely on the CUDA runtime API context anymore in the workflows that use it, and this in turn might break other interop use cases that currently "just work" for SYCL+CUDA interop applications that use the runtime API.

So I think the way to support this is to have the SYCL queue or context created from a native CUstream or CUcontext independent of the rest of CUstreams or CUcontexts, so if there is a pool, there would be kept separate from those. The high-level scheduling based on dependencies could still apply as normal.

AerialMantis commented:
@illuhad @gmlueck @JackAKirk @hdelan FYI I've had to open a new PR for this as I ran into some issues rebasing this, coming back to it after some time. the latest version is now in https://github.com/KhronosGroup/SYCL-Docs/pull/420.

AerialMantis commented:
Closing this as a new PR has been opened for this; https://github.com/KhronosGroup/SYCL-Docs/pull/420.
